cpp/src/glm/ridge_mg.cu

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <common/cumlHandle.hpp>
+#include <common/cuml_comms_int.hpp>
+#include <common/device_buffer.hpp>
+#include <cuda_utils.cuh>
+#include <cuml/common/cuml_allocator.hpp>
+#include <cuml/linear_model/ridge_mg.hpp>
+#include <linalg/add.cuh>
+#include <linalg/gemm.cuh>
+#include <matrix/math.cuh>
+#include <matrix/matrix.cuh>
+#include <opg/linalg/mv_aTb.hpp>
+#include <opg/linalg/svd.hpp>
+#include <opg/preprocess.hpp>
+#include <opg/stats/mean.hpp>
+
+using namespace MLCommon;
+
+namespace ML {
+namespace Ridge {
+namespace opg {
+
+template <typename T>
+void ridgeSolve(const cumlHandle &handle, T *S, T *V,
+                std::vector<Matrix::Data<T> *> &U,
+                const Matrix::PartDescriptor &UDesc,
+                const std::vector<Matrix::Data<T> *> &b, const T *alpha,
+                const int n_alpha, T *w, cudaStream_t *streams, int n_streams,
+                bool verbose) {
+  auto cublasH = handle.getImpl().getCublasHandle();
+  auto cusolverH = handle.getImpl().getcusolverDnHandle();
+  const MLCommon::cumlCommunicator &comm = handle.getImpl().getCommunicator();
+  const std::shared_ptr<deviceAllocator> allocator =
+    handle.getImpl().getDeviceAllocator();
+
+  // Implements this: w = V * inv(S^2 + λ*I) * S * U^T * b
+  T *S_nnz;
+  T alp = T(1);
+  T beta = T(0);
+  T thres = T(1e-10);
+
+  Matrix::setSmallValuesZero(S, UDesc.N, streams[0], thres);
+  allocate(S_nnz, UDesc.N, true);
```

Comment:
You could do it in this PR or create a Github issue for it, but this should be using `device_buffer`

Reply:
I'd vote for this as a follow-up so the first PR is really minimal change.